File: golang-devildyw/middle/L56合并区间.go
package main

import "sort"

func main() {

}

func merge(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}
	// 这题我会啊 先排序
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	result := make([][]int, 0, len(intervals))
	start := 0
	for i := 1; i < len(intervals); i++ {
		if intervals[i-1][1] >= intervals[i][0] {
			intervals[i][1] = max(intervals[i-1][1], intervals[i][1])
			continue
		} else {
			// 区间截止到当前位置-1
			result = append(result, []int{intervals[start][0], intervals[i-1][1]})
			start = i
		}
	}

	result = append(result, []int{intervals[start][0], intervals[len(intervals)-1][1]})

	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L56合并区间.java
package top.dyw.leetcode.Hot100;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=56 lang=java
 *
 * [56] 合并区间
 */

// @lc code=start
public class L56合并区间 {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) ->{
            return a[0] - b[0];
        });
        
        List<int[]> result = new ArrayList<>();
        int left = intervals[0][0];
        int rightMax = intervals[0][1];
        for (int i=0; i<intervals.length; i++) {
            if (rightMax >= intervals[i][0]) {
                rightMax = Math.max(rightMax, intervals[i][1]);
            } else {
                result.add(new int[]{left, rightMax});
                left=intervals[i][0];
                rightMax=intervals[i][1];
            }
        }

        result.add(new int[]{left, rightMax});


        return result.toArray(new int[result.size()][]);
    }
}
// @lc code=end
